stepconf/stepconf.go

```diff
@@ -171,6 +187,260 @@ func setField(field reflect.Value, value, constraint string) error {
+			return fmt.Errorf("minimum value %f is equal to maximum %f, for this case use optional value", minFloat, maxFloat)
+		}
+	}
+
+	if min == nil {
+		return maxErr
+	} else if max == nil {
+		return minErr
+	}
+	if minErr != nil || maxErr != nil {
+		return fmt.Errorf("value %f is out of range %f-%f", value, minFloat, maxFloat)
+	}
+	return nil
+}
+
+func validateRangeFieldTypes(min interface{}, max interface{}, value interface{}) error {
+	if value == nil {
+		return fmt.Errorf("value cannot be nil")
+	}
+	var minType string
+	var maxType string
+	var valueType string
+	var err error
+
+	if valueType, err = getTypeOf(value); err != nil {
+		return err
+	}
+	if min != nil {
+		if minType, err = getTypeOf(min); err != nil {
+			return err
+		}
+	}
+	if max != nil {
+		if maxType, err = getTypeOf(max); err != nil {
+			return err
+		}
+	}
+
+	if maxType != "" && minType != "" && !hasSameContent(minType, maxType, valueType) {
+		return fmt.Errorf("invalid constraint and value combination, minimum is %s, maximum is %s, value is %s, but they should be the same", minType, maxType, valueType)
+	}
+
+	if maxType != "" && !hasSameContent(maxType, valueType) {
+		return fmt.Errorf("invalid constraint and value combination, maximum is %s, value is %s, but they should be the same", maxType, valueType)
+	}
+
+	if minType != "" && !hasSameContent(minType, valueType) {
+		return fmt.Errorf("invalid constraint and value combination, minimum is %s, value is %s, but they should be the same", minType, valueType)
+	}
+	return nil
+}
+
+func hasSameContent(strs ...string) bool {
+	length := len(strs)
+	if length == 1 {
+		return true
+	}
+	firstItem := strs[0]
+	for i := 1; i < length; i++ {
+		if strings.Compare(firstItem, strs[i]) != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func getTypeOf(v interface{}) (string, error) {
+	switch v.(type) {
+	case int64:
+		return "int64", nil
+	case float64:
+		return "float64", nil
+	default:
+		return "unknown", fmt.Errorf("could not find type for %v", v)
+	}
+}
+
+func parseValueStr(value string) (interface{}, error) {
+	var err error
+	var parsedInt int64
+	var parsedFloat float64
+	if parsedInt, err = strconv.ParseInt(value, 10, 64); err != nil {
+		// Could be float
+		if parsedFloat, err = strconv.ParseFloat(value, 64); err != nil {
+			// It is invalid.
+			return nil, fmt.Errorf("value %s is could not be parsed", value)
+		}
+		return parsedFloat, nil
+	}
+	return parsedInt, nil
+}
+
+func getFloatValue(value interface{}) (float64, error) {
+	switch i := value.(type) {
+	case int64:
+		return float64(i), nil
+	case float64:
+		return i, nil
+	case string:
+		var parsedValue interface{}
+		var err error
+		if parsedValue, err = parseValueStr(i); err != nil {
+			return 0, err
+		}
+		return getFloatValue(parsedValue)
+	default:
+		return 0, fmt.Errorf("not supported type %T", value)
+	}
+}
+
+func validateRangeMinFieldValue(min float64, value float64, inclusive bool) error {
+	if inclusive {
+		if min > value {
```

Comment:
please merge the inner ifs into the main ones

Reply:
Merged.